fork-off/src/main.rs

```diff
@@ -12,168 +19,336 @@ pub struct Config {
+        .json()
+        .await
+        .expect("Could not deserialize response as JSON");
+    let result = response["result"].take();
+    serde_json::from_value(result).expect("Incompatible type of the result")
+}
 
-    let mut fork_spec: Value = serde_json::from_str(
-        &fs::read_to_string(&fork_spec_path).expect("Could not read chainspec file"),
-    )?;
+fn construct_json_body(method_name: &str, params: Value) -> Value {
+    serde_json::json!({
+        "jsonrpc": "2.0",
+        "id": 1,
+        "method": method_name,
+        "params": params
+    })
+}
 
-    let hashed_prefixes = prefixes
-        .iter()
-        .map(|prefix| {
-            let hash = format!("0x{}", prefix_as_hex(prefix));
-            info!(target: "fork", "prefix: {}, hash: {}", prefix, hash);
-            hash
-        })
-        .chain([format!("0x{}", hex::encode(":code"))])
-        .collect::<Vec<String>>();
+fn block_hash_body(block_num: Option<u32>) -> Value {
+    let params = serde_json::json!([block_num]);
+    construct_json_body("chain_getBlockHash", params)
+}
 
-    let storage = get_chain_state(&http_rpc_endpoint, &hashed_prefixes).await;
+fn get_keys_paged_body(
+    prefix: StorageKey,
+    count: u32,
+    start_key: Option<StorageKey>,
+    block_hash: Option<BlockHash>,
+) -> Value {
+    let params = serde_json::json!([prefix, count, start_key, block_hash]);
+    construct_json_body("state_getKeysPaged", params)
+}
 
-    info!("Succesfully retrieved chain state {:?}", storage);
+fn get_storage_value_body(key: StorageKey, block_hash: Option<BlockHash>) -> Value {
+    let params = serde_json::json!([key, block_hash]);
+    construct_json_body("state_getStorage", params)
+}
 
-    storage.into_iter().for_each(|(key, value)| {
-        info!(target: "fork","Moving {} to the fork", key);
-        fork_spec["genesis"]["raw"]["top"][key] = value.into();
-    });
+struct StateFetcher {
+    client: Client,
+    http_rpc_endpoint: String,
+}
 
-    // write out the fork spec
-    let json = serde_json::to_string_pretty(&fork_spec)?;
-    info!(target: "fork", "Writing forked chain spec to {}", &write_to_path);
-    write_to_file(write_to_path, json.as_bytes());
+impl StateFetcher {
+    fn new(http_rpc_endpoint: String) -> Self {
+        StateFetcher {
+            client: Client::new(),
+            http_rpc_endpoint,
+        }
+    }
 
-    info!("Done!");
-    Ok(())
-}
+    async fn value_fetching_worker(
+        http_rpc_endpoint: String,
+        id: usize,
+        block: BlockHash,
+        input: Arc<Mutex<Vec<StorageKey>>>,
+        output: Arc<Mutex<Storage>>,
+    ) {
+        let fetcher = StateFetcher::new(http_rpc_endpoint);
+        loop {
+            let maybe_key: Option<StorageKey> = {
+                let mut keys = input.lock();
+                keys.pop()
+            };
+            if let Some(key) = maybe_key {
+                let value = fetcher.get_value(key.clone(), block.clone()).await;
```

Comment:
nit: to reduce nesting:
```suggestion
            let key = match maybe_key {
                Some(key) => key,
                None => break;
            };
```

Reply:
Reduced :)